compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -55,27 +70,73 @@ fun PromisedValue.discard(): MaterialValue {
     materialize()
     if (type !== Type.VOID_TYPE)
         AsmUtil.pop(mv, type)
-    return MaterialValue(mv, Type.VOID_TYPE)
+    return codegen.voidValue
+}
+
+private val IrType.unboxed: IrType
+    get() = InlineClassAbi.getUnderlyingType(erasedUpperBound)
+
+fun PromisedValue.coerceInlineClasses(type: Type, irType: IrType, target: Type, irTarget: IrType): PromisedValue? {
+    val isFromTypeInlineClass = irType.erasedUpperBound.isInline
+    val isToTypeInlineClass = irTarget.erasedUpperBound.isInline
+    if (!isFromTypeInlineClass && !isToTypeInlineClass) return null
+
+    val isFromTypeUnboxed = isFromTypeInlineClass && typeMapper.mapType(irType.unboxed) == type
+    val isToTypeUnboxed = isToTypeInlineClass && typeMapper.mapType(irTarget.unboxed) == target
+
+    return when {
+        isFromTypeUnboxed && !isToTypeUnboxed -> object : PromisedValue(codegen, target, irTarget) {
+            override fun materialize() {
+                this@coerceInlineClasses.materialize()
+                // TODO: This is broken for type parameters
```

Comment:
Yes exactly. Rather than fixing it in the type mapper I changed the code to always go through the erased upper bound for inline class boxing instead.

It would be nice if we could always do this, but unfortunately the type mapper has additional side effects when processing generic types.